### ISSUE TITLE
MiniLLM: right-padding leaks into the reverse-KL advantage, and length normalization depends on batch padding

### ISSUE THREAD

**Issue by behroozazarkhalili** [huggingface/trl#7024]
### Reproduction

Both points below are on `main` and are independent of #6635, which only changes how the discounted sum is computed. They came up while reviewing that PR.

**1. Right-padding is never masked out of the advantage.** `MiniLLMTrainer.compute_loss` builds the mask it hands to `_compute_advantage` from the raw completion ids:

```python
shifted_labels = input_ids[:, prompt_lengths:]
...
mask = shifted_labels != -100
```

`input_ids` never contains `-100`; the `-100` fill is applied to `labels`, a separate tensor built a few lines earlier from `attention_mask`. So `mask` is all ones, padded positions keep a reward of `teacher_logp - student_logp` on the pad token, and every earlier position's advantage sums those padded rewards in. With one 4-slot completion whose last two slots are padding, per-token rewards `teacher_logp - student_logp` of `[1.5, 1.0, 2.0, 2.0]` and `gamma=1.0`, the mask the caller builds and the mask it should build give different advantages at every real position:

```
caller mask       [1, 1, 1, 1]   advantages [6.5, 5.0, 4.0, 2.0]
completion mask   [1, 1, 0, 0]   advantages [2.5, 1.0, 0.0, 0.0]
```

The fix is a one-line change, `shifted_labels = labels[:, prompt_lengths:]`, since `labels` already carries the `-100` fill.

**2. Under `length_normalization=True` the result depends on how much padding the batch carries.** `_compute_advantage` replaces every masked position with `1e-4` before building the denominator, so the denominator of a short completion includes `1e-4 * gamma^(i-t)` for each pad slot after it. With a constant reward of `1.0` and `gamma=1.0`, which should give an advantage of exactly `1.0` everywhere, the value at the first position moves with the padded length:

```
padded to T=512   L=1     first=0.9514   last_valid=0.9514
                  L=128   first=0.9997   last_valid=0.9630
                  L=512   first=1.0000   last_valid=1.0000
```

The same completion gets a different advantage depending on the longest completion in its batch. The `1e-4` exists so that a fully masked row does not divide by zero; a `clamp(min=...)` on the finished denominator would give that guarantee without adding the pad slots to every real position's length.

Both tables come from calling the method directly with a stub `self`, the way `tests/experimental/test_minillm_trainer.py` does:

```python
import torch, types
from trl.experimental.minillm.minillm_trainer import MiniLLMTrainer

student = torch.tensor([[-2.0, -1.5, -3.0, -3.0]])
teacher = torch.tensor([[-0.5, -0.5, -1.0, -1.0]])
stub = types.SimpleNamespace(gamma=1.0, length_normalization=False)
print(MiniLLMTrainer._compute_advantage(stub, student, teacher, torch.ones(1, 4)))
print(MiniLLMTrainer._compute_advantage(stub, student, teacher, torch.tensor([[1.0, 1.0, 0.0, 0.0]])))

stub = types.SimpleNamespace(gamma=1.0, length_normalization=True)
for length in (1, 128, 512):
    mask = torch.zeros(1, 512); mask[0, :length] = 1
    adv = MiniLLMTrainer._compute_advantage(stub, torch.zeros(1, 512), torch.ones(1, 512), mask)[0]
    print(length, adv[0].item(), adv[length - 1].item())
```

### System Info

- Platform: Linux 5.14.0 (x86_64)
- Python 3.11
- TRL: 1.10.0 (dev, `main`)
- PyTorch 2.11.0
- Transformers 5.11.0
- Accelerate 1.13.0

### Checklist

- [x] I have checked that my issue isn't already filed
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible
- [x] Any code provided is properly formatted in code blocks
- [x] Any traceback provided is complete

**Comment by daniellaah** [comment]
Thanks for the detailed report! I reproduced both issues on the current main branch. Are you planning to submit a fix, or would it be okay if I worked on a PR for this?

**Comment by behroozazarkhalili** [comment]
Hi,
Thanks.

I'll submit the fix soon this week.


Best Regards,
Behrooz Azarkhalili
*---------------------------------*

**Comment by Adya-Prasad** [comment]
Hey @behroozazarkhalili, I want to start contributing to TRL and going through some issues; I noticed that you are handling quite a few. Could you suggest an open issue for me to get started with (difficulty level: easy to medium)?
_Context: I have contributed to PEFT, so I can pick related to that too_

**Comment by behroozazarkhalili** [comment]
Hi, thanks for the interest. This one is already fixed in #7044, so it is not a good starting point. I am not a maintainer, so I cannot assign issues; the usual path here is the `good first issue` and `good second issue` labels, then a comment on the issue saying you are taking it so the maintainers can confirm nobody else is on it. Issues where @qgallouedec or @albertvillanova have already stated what they want done tend to go smoothest.

**Comment by Adya-Prasad** [comment]
Got it and thank you. But pls keep in touch

**Comment by behroozazarkhalili** [comment]
Closing this myself. I found it by reading the code, not from a failure in a real training run, and per #6862 that isn't the bar here. Happy to reopen if someone hits it in practice.
